platform/ios/MapLibre.docc/LocationPrivacyExample.md

```diff
@@ -40,75 +40,81 @@ enum LocationAccuracyState {
+    @Published var locationAccuracy: LocationAccuracyState = .unknown
+    @Published var showTemporaryLocationAuthorization = false
+}
 
-    func mapView(_: MLNMapView, didChangeLocationManagerAuthorization manager: MLNLocationManager) {
+class PrivacyExampleCoordinator: NSObject, MLNMapViewDelegate {
+    @ObservedObject private var mapViewModel: PrivacyExampleViewModel
+    private var pannedToUserLocation = false
+    
+    init(mapViewModel: PrivacyExampleViewModel) {
+        self.mapViewModel = mapViewModel
+        super.init()
+    }
+    
+    @MainActor func mapView(_: MLNMapView, didChangeLocationManagerAuthorization manager: MLNLocationManager) {
         guard let accuracySetting = manager.accuracyAuthorization else {
             return
         }
-
+        
         switch accuracySetting() {
         case .fullAccuracy:
-            locationAccuracy = .fullAccuracy
+            mapViewModel.locationAccuracy = .fullAccuracy
         case .reducedAccuracy:
-            locationAccuracy = .reducedAccuracy
+            mapViewModel.locationAccuracy = .reducedAccuracy
         @unknown default:
-            locationAccuracy = .unknown
+            mapViewModel.locationAccuracy = .unknown
         }
     }
 
     // when a location is available for the first time, we fly to it
     func mapView(_ mapView: MLNMapView, didUpdate _: MLNUserLocation?) {
-        if pannedToUserLocation {
-            return
-        }
+        guard !pannedToUserLocation else { return }
         guard let userLocation = mapView.userLocation else {
             print("User location is currently not available.")
             return
         }
         mapView.fly(to: MLNMapCamera(lookingAtCenter: userLocation.coordinate, altitude: 100_000, pitch: 0, heading: 0))
+        pannedToUserLocation = true
     }
 }
 
-struct LocationPrivacyExample: UIViewRepresentable {
-    @Binding var mapView: MLNMapView
-    @Binding var locationAccuracy: LocationAccuracyState
-
-    func makeCoordinator() -> Coordinator {
-        Coordinator(mapView: $mapView, locationAccuracy: $locationAccuracy)
+struct PrivacyExampleRepresentable: UIViewRepresentable {
+    @ObservedObject var mapViewModel: PrivacyExampleViewModel
+    
+    func makeCoordinator() -> PrivacyExampleCoordinator {
+        PrivacyExampleCoordinator(mapViewModel: mapViewModel)
     }
 
     func makeUIView(context: Context) -> MLNMapView {
         let mapView = MLNMapView()
-        mapView.showsUserLocation = true
         mapView.delegate = context.coordinator
-
+        mapView.showsUserLocation = true
         return mapView
     }
 
-    func updateUIView(_: MLNMapView, context _: Context) {}
+    func updateUIView(_ mapView: MLNMapView, context _: Context) {
+        if mapViewModel.showTemporaryLocationAuthorization {
+            let purposeKey = "MLNAccuracyAuthorizationDescription"
+            mapView.locationManager.requestTemporaryFullAccuracyAuthorization?(withPurposeKey: purposeKey)
```

Comment:
Nothing sets showTemporaryLocationAuthorization to false again, so any update will call that request function again once set to true. According to https://developer.apple.com/documentation/corelocation/cllocationmanager/3600217-requesttemporaryfullaccuracyauth there is a completion function, I think we should use that and set it back to false?

Reply:
That is part of the problem I have with `UIViewRepresentable` 

We have a published prop `showTemporaryLocationAuthorization` that will trigger a `updateUIView` update, but then inside `updateUIView` if we yet set `showTemporaryLocationAuthorization` back to false, it technically triggers another `updateUIView`. Now in this case, we have a bool we can check, but still. It just feels "meh" to me.

So let's take this as an example of setting `showTemporaryLocationAuthorization` back to false:

```
func updateUIView(_ mapView: MLNMapView, context _: Context) {
        if mapViewModel.showTemporaryLocationAuthorization {
            let purposeKey = "MLNAccuracyAuthorizationDescription"
            mapView.locationManager.requestTemporaryFullAccuracyAuthorization?(withPurposeKey: purposeKey)
            mapViewModel.showTemporaryLocationAuthorization = false
        }
    }
```
When try that, we even get a warning about it:

<img width="1530" alt="tears" src="https://github.com/maplibre/maplibre-native/assets/2292621/153bea6d-a0aa-47b1-9f69-90141097ac66">

I have read so people dealing and fighting with `UIViewRepresentable` like this, examples:

1. https://stackoverflow.com/a/76919202
2. https://www.reddit.com/r/SwiftUI/comments/kti9r9/uiviewrepresentable_how_to_update_bindings/
3. there are millions more I won't link to

So to fix that warning people are wrapping it in a dispatch to main, which I don't like personally:

```
    func updateUIView(_ mapView: MLNMapView, context _: Context) {
        if mapViewModel.showTemporaryLocationAuthorization {
            let purposeKey = "MLNAccuracyAuthorizationDescription"
            mapView.locationManager.requestTemporaryFullAccuracyAuthorization?(withPurposeKey: purposeKey)
            DispatchQueue.main.async {
                mapViewModel.showTemporaryLocationAuthorization = false
            }
        }
    }
```


And also, to your comment about `CLLocationManager` having a completion callback, this example uses `MLNLocationManager` which wraps `CLLocatonManager`, and maplbire MLNLocationManager's version of  `requestTemporaryFullAccuracyAuthorization` hasn't been updated to support the completion callback from what I can tell... unless I'm just missing it for some reason... can you double check please?

Also, this brings up one of your points in  our [Slack discussion](https://osmus.slack.com/archives/C06U5MM097B/p1715614940371019?thread_ts=1715612553.323529&cid=C06U5MM097B) about most apps they have their own Location Manager functionality, but I think @louwers intention was to keep these examples small and to the point and adding yet another class / code to handle location manager increases the complexity of the example.

Reply:
But to your comment regarding `I think we should use that and set it back to false?`, I agree 💯  ... but what am I missing with `UIViewRepresentable`?

This is why In my app I do not use `updateUIView`, but instead keep a weak reference to the underlying view. I know that seems worse, but it actually makes it so much cleaner.

For example, [look at this where people are using MKMapView](https://stackoverflow.com/a/73041182) and their implementation removes/adds annotations each update!

I mean that will work, but that seems highly unnecessary. 

And in most large apps we are going to have so many computed props and bindings that need to interface with `MLNMapView`, imagine how much unnecessary work will be done.

Thoughts? Suggestions? 

😄

Reply:
More links:

1. https://stackoverflow.com/questions/58460847/calling-a-local-uiviewcontroller-function-from-a-swiftui-view#comment113325151_61099536
2. https://stackoverflow.com/questions/58039286/how-do-you-call-a-method-on-a-uiview-from-outside-the-uiviewrepresentable-in-swi
3. https://stackoverflow.com/questions/56584059/send-tapaction-from-swiftui-button-action-to-uiview-function

Reply:
This "inefficient" recreation is completely normal in SwiftUI. At the end of the day, it doesn't matter if update is called a few times - per runloop iteration its grouped together, so will only be called once - and as long as the recreation takes microseconds, it will have no impact on users. I know at every developers heart is efficiency - it took me quite a while to get used to this new paradigm too, but SwiftUI takes the "premature optimization is the root of all evil" approach: don't worry about it until you actually have a perceivable impact on performance, and if you ever do, then you can start optimizing - in our case this will have no perceivable impact.

But you are right, we need to remove that "Publishing changes from within view updates is not allowed, this will cause undefined behavior." issue. I assume that if we did have access to the "completion" version of requestTemporaryFullAccuracyAuthorization and call showTemporaryLocationAuthorization = false there, this error would not appear. So we either need to expose that function or, if we're looking for a quick and dirty solution, then yes we need to do the workaround with DispatchQueue.main.async.

As mentioned, I use my own location manager instead of maplibre's - a modified version of https://github.com/malcommac/SwiftLocation, this allows me to create this code for permission requests:

```Swift
 MapView(styleURL: styleURL, camera: $camera) // this is based on maplibre-swiftui-dsl-playground
            .task {
                for await event in await locationManager.startMonitoringAuthorization() {
                    print("Authorization status did change: \(event.authorizationStatus)")
                    // break to interrupt the stream
                    if event.authorizationStatus == .authorizedWhenInUse {
                        showUserLocation = true
                    } else {
                        showUserLocation = false
                    }
                }
            }
            .task {
                showUserLocation = locationManager.authorizationStatus == .authorizedWhenInUse
            }
```

And this is then a button to get the current location:

```Swift
        Button {
            Task {
                do {
                    location.accuracy = .threeKilometers
                    try await location.requestPermission(.whenInUse) // obtain the permissions
                    let userLocation = try await location.requestLocation()
                    
                    if let coordinates = userLocation.location?.coordinate {
                        withAnimation {
                            camera = MapViewCamera.center(coordinates, zoom: 17)
                            
                        }
                    } else {
                        print("location error: got no coordinates")
                    }
                } catch {
                    print("location error: \(error)")
                }
            }
        } label: {
            Image(systemName: "location")
        }
```
The code is clean, uses async/await, and changes "showUserLocation" based on changes to permissions - with permissions its on, without permissions its off and does not trigger a permission request.

I'm not suggesting we use https://github.com/malcommac/SwiftLocation in the official maplibre examples, but if we have an example, we need to decide if its supposed to be a "best practice" example or a "quick and dirty get it up and running" example. Personally I'd argue for "best practice, even if more complicated" because people will copy and paste our examples into their code and location management is not an easy task. But I do understand that complicated examples could scare off devs too. 🤷‍♂️

Reply:
> Personally I'd argue for "best practice, even if more complicated" because people will copy and paste our examples into their code and location management is not an easy task.

Fully agreed. Even an example with an external dependency would be fine by me. Let's not forget the iOS MapLibre codebase is quite rusty, so it should not be surprising some things are not very modern anymore.

I'm just transplanting some examples from https://github.com/mapbox/ios-sdk-examples, I'm not a veteran iOS developer like you, so I am happy for your critical eye!